<a href="https://colab.research.google.com/github/luislinno/Projeto_Predicao_Diabetes/blob/main/Classificacao_Diabtes.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# APRENDIZAGEM SUPERVISIONADA: CLASSIFICAÇÃO

Este projeto tem por objetivo desenvolver um algoritmo de Machine Learning para prever a tendência de uma pessoa desenvolver algum tipo de doença diabética com base em alguns fatores clínicos e laboratoriais.

# PRÉ-PROCESSAMENTO

In [4]:
# Import das biblicotes

import numpy as np
import pandas as pd

In [5]:
#Improtação do dataset tratado. 

df = pd.read_csv('/content/drive/MyDrive/Arquivos/diabetes_tratado.csv',
                sep=',', encoding='utf-8')

In [6]:
df.head()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6    148.0           72.0           35.0    156.0  33.6   
1            1     85.0           66.0           29.0    156.0  26.6   
2            8    183.0           64.0           29.0    156.0  23.3   
3            1     89.0           66.0           23.0     94.0  28.1   
4            0    137.0           40.0           35.0    168.0  43.1   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1

In [7]:
df.shape

(768, 9)

# ATRIBUTOS PREVISORES E ALVO

In [8]:
#iloc = Localiza pelo indice

#: -> Todos os valores

# 0:N:Intervalo de coluna

previsores = df.iloc[:, 0:8].values

In [9]:
previsores

array([[  6.   , 148.   ,  72.   , ...,  33.6  ,   0.627,  50.   ],
       [  1.   ,  85.   ,  66.   , ...,  26.6  ,   0.351,  31.   ],
       [  8.   , 183.   ,  64.   , ...,  23.3  ,   0.672,  32.   ],
       ...,
       [  5.   , 121.   ,  72.   , ...,  26.2  ,   0.245,  30.   ],
       [  1.   , 126.   ,  60.   , ...,  30.1  ,   0.349,  47.   ],
       [  1.   ,  93.   ,  70.   , ...,  30.4  ,   0.315,  23.   ]])

In [10]:
previsores.shape

(768, 8)

In [11]:
alvo = df.iloc[:, 8].values

In [12]:
# Atributo alvo.

alvo

array([1, 0, 1, 0, 1, 0, 1, 0, 1, 1, 0, 1, 0, 1, 1, 1, 1, 1, 0, 1, 0, 0,
       1, 1, 1, 1, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0, 0, 1,
       0, 1, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 1, 0,
       1, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0,
       1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1,
       1, 1, 0, 0, 1, 1, 1, 0, 0, 0, 1, 0, 0, 0, 1, 1, 0, 0, 1, 1, 1, 1,
       1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0,
       1, 1, 0, 0, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 1,
       0, 1, 0, 1, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 0, 0, 1, 1, 0, 1, 0, 1,
       1, 1, 0, 0, 0, 0, 0, 0, 1, 1, 0, 1, 0, 0, 0, 1, 1, 1, 1, 0, 1, 1,
       1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 0, 1, 1, 1, 1, 0, 0, 0,
       1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 1, 0, 1, 0, 0,
       1, 0, 1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 1, 1, 0,
       0, 1, 0, 0, 0, 1, 1, 1, 0, 0, 1, 0, 1, 0, 1,

In [13]:
alvo.shape

(768,)

# VERIFICANDO AS MELHORES FEATURES

In [14]:
from sklearn.feature_selection import SelectKBest

In [15]:
# Escolhendo as melhores features com Kbest

features_list = ('Pregnancies', 'Glucose', 'BloodPressure', 'SkinThickness', 'Insulin', 'BMI', 'DiabetesPedigreeFunction', 'Age')

k_best_features = SelectKBest(k='all')
k_best_features.fit_transform(previsores, alvo)
k_best_features_score = k_best_features.scores_
raw_pairs = zip(features_list[1:], k_best_features_score)
ordered_pairs = list(reversed(sorted(raw_pairs, key=lambda x: x[1])))

k_best_features_final = dict(ordered_pairs[:8])
best_feature = k_best_features_final.keys()
print('')
print('Melhores Features: ')
print(k_best_features_final)


Melhores Features: 
{'BloodPressure': 245.85920854104933, 'DiabetesPedigreeFunction': 82.56242947356262, 'Glucose': 39.670227393616116, 'Insulin': 37.07853796802245, 'BMI': 36.950492613849946, 'Age': 23.871300204344593, 'SkinThickness': 21.631579805760936}


# ANÁLISE DAS ESCALAS DOS ATRIBUTOS (ESCALONAMENTO)

In [16]:
df.describe()

Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
count   768.000000  768.000000     768.000000     768.000000  768.000000   
mean      3.845052  121.686763      72.386719      29.108073  155.768229   
std       3.369578   30.435949      12.096642       8.791221   85.021408   
min       0.000000   44.000000      24.000000       7.000000   14.000000   
25%       1.000000   99.750000      64.000000      25.000000  121.500000   
50%       3.000000  117.000000      72.000000      29.000000  156.000000   
75%       6.000000  140.250000      80.000000      32.000000  156.000000   
max      17.000000  199.000000     122.000000      99.000000  846.000000   

              BMI  DiabetesPedigreeFunction         Age     Outcome  
count  768.000000                768.000000  768.000000  768.000000  
mean    32.457464                  0.471876   33.240885    0.348958  
std      6.875151                  0.331329   11.760232    0.476951  
min     18.200000                  0.078000   21.000000    0.000000  
25%     27.500000                  0.243750   24.000000    0.000000  
50%     32.400000                  0.372500   29.000000    0.000000  
75%     36.600000                  0.626250   41.000000    1.000000  
max     67.100000                  2.420000   81.000000    1.000000

Padronização (utilizando a média e o desvio padrão como referência). Normalização (utiliza os valores máximo e mínimo como referência).

In [17]:
from sklearn.preprocessing import StandardScaler

In [18]:
# Escalonamento dos atributos. 

previsores_esc = StandardScaler().fit_transform(previsores)

In [19]:
previsores_esc

array([[ 0.63994726,  0.86510807, -0.03198993, ...,  0.16629174,
         0.46849198,  1.4259954 ],
       [-0.84488505, -1.20616153, -0.5283186 , ..., -0.85253118,
        -0.36506078, -0.19067191],
       [ 1.23388019,  2.0158134 , -0.69376149, ..., -1.33283341,
         0.60439732, -0.10558415],
       ...,
       [ 0.3429808 , -0.0225789 , -0.03198993, ..., -0.91074963,
        -0.68519336, -0.27575966],
       [-0.84488505,  0.14180757, -1.02464727, ..., -0.34311972,
        -0.37110101,  1.17073215],
       [-0.84488505, -0.94314317, -0.19743282, ..., -0.29945588,
        -0.47378505, -0.87137393]])

In [20]:
previsoresdf = pd.DataFrame(previsores_esc)
previsoresdf

0         1         2         3         4         5         6  \
0    0.639947  0.865108 -0.031990  0.670643  0.002728  0.166292  0.468492   
1   -0.844885 -1.206162 -0.528319 -0.012301  0.002728 -0.852531 -0.365061   
2    1.233880  2.015813 -0.693761 -0.012301  0.002728 -1.332833  0.604397   
3   -0.844885 -1.074652 -0.528319 -0.695245 -0.726976 -0.634212 -0.920763   
4   -1.141852  0.503458 -2.679076  0.670643  0.143961  1.548980  5.484909   
..        ...       ...       ...       ...       ...       ...       ...   
763  1.827813 -0.680125  0.298896  2.150354  0.285194  0.064409 -0.908682   
764 -0.547919  0.010298 -0.197433 -0.239949  0.002728  0.632039 -0.398282   
765  0.342981 -0.022579 -0.031990 -0.695245 -0.515126 -0.910750 -0.685193   
766 -0.844885  0.141808 -1.024647 -0.012301  0.002728 -0.343120 -0.371101   
767 -0.844885 -0.943143 -0.197433  0.215347  0.002728 -0.299456 -0.473785   

            7  
0    1.425995  
1   -0.190672  
2   -0.105584  
3   -1.041549  
4   -0.020496  
..        ...  
763  2.532136  
764 -0.531023  
765 -0.275760  
766  1.170732  
767 -0.871374  

[768 rows x 8 columns]

In [21]:
previsoresdf.describe()

0             1             2             3             4  \
count  7.680000e+02  7.680000e+02  7.680000e+02  7.680000e+02  7.680000e+02   
mean  -6.476301e-17 -3.561966e-16  5.782412e-18 -1.526557e-16  1.272131e-16   
std    1.000652e+00  1.000652e+00  1.000652e+00  1.000652e+00  1.000652e+00   
min   -1.141852e+00 -2.554131e+00 -4.002619e+00 -2.516429e+00 -1.668528e+00   
25%   -8.448851e-01 -7.212214e-01 -6.937615e-01 -4.675972e-01 -4.033168e-01   
50%   -2.509521e-01 -1.540881e-01 -3.198993e-02 -1.230129e-02  2.727806e-03   
75%    6.399473e-01  6.103090e-01  6.297816e-01  3.291706e-01  2.727806e-03   
max    3.906578e+00  2.541850e+00  4.104082e+00  7.955377e+00  8.123620e+00   

                  5             6             7  
count  7.680000e+02  7.680000e+02  7.680000e+02  
mean   3.515706e-16  2.451743e-16  1.931325e-16  
std    1.000652e+00  1.000652e+00  1.000652e+00  
min   -2.075119e+00 -1.189553e+00 -1.041549e+00  
25%   -7.215397e-01 -6.889685e-01 -7.862862e-01  
50%   -8.363615e-03 -3.001282e-01 -3.608474e-01  
75%    6.029301e-01  4.662269e-01  6.602056e-01  
max    5.042087e+00  5.883565e+00  4.063716e+00

# RESUMO PRÉ-PROCESSAMENTO

alvo = variavel que se pretende atingir (tem ou não doença cardíaca).

previsores = conjunto de variaveis previsoras; 

previsores_esc = conjunto de variváveis previsoreas escalonada.

# BASE DE TREINO E TESTE

In [22]:
from sklearn.model_selection import train_test_split

Paramêtros train_test_split: >Arrays: nomes dos atributos previsores e alvo. >test_size: tamanho em porcentagem dos dados de teste. default é none. >train_size: tamanho em porcentagem dos dados de treinamento. default é none. >random_state: noe=meação de um estado aleatório. >shuffle: embaralhamento dos dados aleatórios. Associados com o random_state ocorre o mesmo embaralhamento sempre. Defaul é True. >stratify: Possibilidade de dividir os dados de forma estratificada. Default é None (Nesse caso é mantido a proporção, isto é, se tem 30% de zeros e 70% de 1 no dataframe, na separação em treinamento e teste se manterá esse proporção). 

In [326]:
x_treino, x_teste, y_treino, y_teste = train_test_split(previsores, alvo, test_size= 0.2, random_state = 0)

In [24]:
x_treino.shape

(614, 8)

In [25]:
x_teste.shape

(154, 8)

In [26]:
y_treino.shape

(614,)

In [27]:
y_teste.shape

(154,)

# NAIVE BAYES

### Treinamento Algoritmo

In [28]:
from sklearn.naive_bayes import GaussianNB

In [29]:
naive = GaussianNB()
naive.fit(x_treino, y_treino)

GaussianNB()

In [30]:
previsoes_naive = naive.predict(x_teste)
previsoes_naive

array([1, 0, 0, 1, 0, 0, 1, 1, 1, 0, 1, 1, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0,
       0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 0, 0, 0, 1,
       1, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 1, 1,
       1, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0,
       1, 0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 1,
       0, 0, 1, 1, 1, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0])

In [31]:
y_teste

array([1, 0, 0, 1, 0, 0, 1, 1, 0, 0, 1, 1, 0, 0, 0, 0, 1, 0, 0, 0, 1, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 1,
       1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 0, 0, 1, 1, 1, 1, 0, 0, 1, 1, 1,
       1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0,
       1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1,
       0, 1, 1, 1, 1, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0,
       0, 1, 0, 1, 0, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 1, 0, 0])

In [32]:
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

In [33]:
print('Acurácia: %.2f%%' % (accuracy_score(y_teste, previsoes_naive) * 100.0))

Acurácia: 78.57%


In [34]:
confusion_matrix(y_teste, previsoes_naive)

array([[92, 15],
       [18, 29]])

In [35]:
print(classification_report(y_teste, previsoes_naive))

              precision    recall  f1-score   support

           0       0.84      0.86      0.85       107
           1       0.66      0.62      0.64        47

    accuracy                           0.79       154
   macro avg       0.75      0.74      0.74       154
weighted avg       0.78      0.79      0.78       154



### Análise dados de treino

In [36]:
previsores_treino = naive.predict(x_treino)
previsores_treino

array([1, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1,
       0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 1, 1, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 1, 1, 0,
       0, 0, 0, 1, 0, 1, 0, 1, 0, 1, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0,
       1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 1, 1, 0, 0, 1, 0, 1, 1, 0, 0, 1, 0, 1, 0, 0, 1, 1, 0, 0,
       0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1,
       1, 1, 1, 0, 0, 0, 0, 0, 1, 0, 1, 0, 1, 1, 1, 0, 0, 1, 1, 1, 0, 1,
       0, 0, 0, 1, 1, 0, 0, 0, 1, 1, 1, 0, 1, 0, 1, 0, 0, 1, 1, 0, 0, 0,
       0, 1, 0, 0, 0, 1, 1, 0, 0, 1, 1, 0, 1, 0, 1, 0, 0, 0, 1, 0, 1, 0,
       1, 0, 0, 0, 1, 1, 0, 1, 1, 0, 0, 0, 1, 0, 0, 1, 0, 1, 0, 0, 0, 1,
       0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 1, 1, 0, 1, 0, 0, 1,
       0, 0, 1, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0,

In [37]:
print('Acurácia: %.2f%%' % (accuracy_score(y_treino, previsores_treino) * 100.0))

Acurácia: 74.92%


In [38]:
confusion_matrix(y_treino, previsores_treino)

array([[324,  69],
       [ 85, 136]])

### Validação Cruzada

In [39]:
from sklearn.model_selection import KFold
from sklearn.model_selection import cross_val_score

In [40]:
#Separando os dados em folds

kfold = KFold(n_splits = 30, shuffle=True, random_state =5)

In [41]:
#Criando o modelo

modelo = GaussianNB()
resultado = cross_val_score(modelo, previsores, alvo, cv = kfold)
resultado

array([0.76923077, 0.80769231, 0.76923077, 0.80769231, 0.88461538,
       0.73076923, 0.73076923, 0.69230769, 0.76923077, 0.76923077,
       0.76923077, 0.69230769, 0.76923077, 0.69230769, 0.73076923,
       0.69230769, 0.80769231, 0.80769231, 0.76      , 0.64      ,
       0.76      , 0.68      , 0.76      , 0.76      , 0.76      ,
       0.68      , 0.56      , 0.88      , 0.68      , 0.8       ])

In [42]:
#Usamos a média e o desvio padrão

print('Acurácia Média: %.2f%%' % (resultado.mean() * 100.0))

Acurácia Média: 74.71%


Naive Bayes = 78.57% (treino e teste) - 121 acertos e 74.71% (validação cruzada) - previsores

# MÁQUINAS DE VETORES DE SUPORTE(SVM)

### Treinamento Algoritmo

In [43]:
from sklearn.svm import SVC

In [44]:
svm = SVC(kernel='rbf', random_state=1, C = 9)
svm.fit(x_treino, y_treino)

SVC(C=9, random_state=1)

In [45]:
previsoes_svm = svm.predict(x_teste)
previsoes_svm

array([1, 0, 0, 1, 0, 0, 1, 1, 0, 0, 1, 1, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0,
       0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1,
       1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 1, 1,
       1, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       1, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1,
       0, 0, 1, 0, 1, 1, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0,
       0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0])

In [46]:
y_teste

array([1, 0, 0, 1, 0, 0, 1, 1, 0, 0, 1, 1, 0, 0, 0, 0, 1, 0, 0, 0, 1, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 1,
       1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 0, 0, 1, 1, 1, 1, 0, 0, 1, 1, 1,
       1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0,
       1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1,
       0, 1, 1, 1, 1, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0,
       0, 1, 0, 1, 0, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 1, 0, 0])

In [47]:
print('Acurácia: %.2f%%' % (accuracy_score(y_teste, previsoes_svm) * 100))

Acurácia: 80.52%


In [48]:
confusion_matrix(y_teste, previsoes_svm)

array([[98,  9],
       [21, 26]])

In [49]:
print(classification_report(y_teste, previsoes_svm))

              precision    recall  f1-score   support

           0       0.82      0.92      0.87       107
           1       0.74      0.55      0.63        47

    accuracy                           0.81       154
   macro avg       0.78      0.73      0.75       154
weighted avg       0.80      0.81      0.80       154



### Análise de dados de treino

In [50]:
previsoes_svm_treino = svm.predict(x_treino)
previsoes_svm_treino

array([1, 0, 0, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0,
       0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0,
       0, 0, 0, 1, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 1, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0,
       1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0,
       0, 0, 0, 1, 1, 0, 0, 1, 0, 1, 1, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0, 0,
       0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1,
       1, 1, 1, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 1, 1, 0, 0, 0, 1, 1, 0, 1,
       0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 0,
       0, 1, 0, 0, 0, 1, 1, 0, 0, 1, 1, 0, 1, 0, 1, 0, 0, 0, 1, 0, 1, 0,
       0, 0, 1, 0, 1, 1, 0, 1, 1, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 1,
       0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 1, 0, 1, 0, 0, 0,
       0, 0, 1, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0,

In [51]:
print('Acurácia: %.2f%%' % (accuracy_score(y_treino, previsoes_svm_treino) * 100))

Acurácia: 77.36%


In [52]:
confusion_matrix(y_treino, previsoes_svm_treino)

array([[354,  39],
       [100, 121]])

### Validação Cruzada

In [53]:
#Separando os dados em folds

Kfold = KFold(n_splits = 30, shuffle =True, random_state = 5)

In [337]:
#Criando o modelo 

modelo = SVC(kernel='rbf', random_state=1, C = 9)
resultado = cross_val_score(modelo, previsores, alvo, cv = kfold)

In [338]:
#Usando a média e o desvio padrão

print('Acurária Média: %.2f%%' % (resultado.mean() * 100))

Acurária Média: 76.42%


Naive Bayes = 78.57% (treino e teste) - 121 acertos e 74.71% (validação cruzada) - previsores

Máquinas de Vetores de Suporte (SVM) = 80.52% (Treino e Teste) - 124 acertos e 76.42% (validação cruzada) - previsores. 

# REGRESSÃO LOGÍSTICA

### Treinamento Algoritmo

In [56]:
from sklearn.linear_model import LogisticRegression

In [57]:
#Configurando os parâmetros da biblioteca importada

logistica = LogisticRegression(random_state=1, max_iter=200, penalty='l2', tol=1, C=1, solver='lbfgs')

In [58]:
#Treinamento

logistica.fit(x_treino, y_treino)

LogisticRegression(C=1, max_iter=200, random_state=1, tol=1)

In [59]:
logistica.intercept_

array([-8.47114046])

In [60]:
logistica.coef_

array([[ 0.09710828,  0.03733893, -0.01434469, -0.00447462, -0.00125315,
         0.09746258,  0.7262245 ,  0.01836674]])

In [61]:
previsores_logistica = logistica.predict(x_teste)
previsores_logistica

array([1, 0, 0, 1, 0, 0, 1, 1, 0, 0, 1, 1, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0,
       0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1,
       1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 1, 1,
       1, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       1, 0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1,
       0, 0, 1, 0, 1, 1, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0])

In [62]:
y_teste

array([1, 0, 0, 1, 0, 0, 1, 1, 0, 0, 1, 1, 0, 0, 0, 0, 1, 0, 0, 0, 1, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 1,
       1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 0, 0, 1, 1, 1, 1, 0, 0, 1, 1, 1,
       1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0,
       1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1,
       0, 1, 1, 1, 1, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0,
       0, 1, 0, 1, 0, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 1, 0, 0])

In [63]:
print('Acurácia: %.2f%%' % (accuracy_score(y_teste, previsores_logistica) * 100.0))

Acurácia: 80.52%


In [64]:
confusion_matrix(y_teste, previsores_logistica)

array([[97, 10],
       [20, 27]])

In [65]:
print(classification_report(y_teste, previsores_logistica))

              precision    recall  f1-score   support

           0       0.83      0.91      0.87       107
           1       0.73      0.57      0.64        47

    accuracy                           0.81       154
   macro avg       0.78      0.74      0.75       154
weighted avg       0.80      0.81      0.80       154



### Análise de dados de treino

In [66]:
previsores_treino_logistica = logistica.predict(x_treino)
previsores_treino_logistica

array([1, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0,
       0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0,
       0, 0, 0, 1, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 1, 0, 0,
       0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0,
       1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0,
       0, 0, 0, 1, 1, 0, 0, 1, 0, 1, 1, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0, 1,
       0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 1, 0, 0, 0, 0, 1, 0, 1, 1, 0, 0, 1,
       1, 1, 1, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 1, 1, 0, 0, 0, 1, 1, 0, 0,
       0, 0, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 1, 1, 0, 0, 1, 1, 0, 0, 0,
       0, 1, 0, 0, 0, 1, 1, 0, 0, 1, 1, 0, 1, 0, 1, 0, 0, 0, 1, 0, 1, 0,
       1, 0, 1, 0, 1, 1, 0, 1, 1, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 1,
       0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 1, 1, 0, 1, 0, 0, 0,
       0, 0, 1, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0,

In [67]:
y_treino

array([1, 0, 0, 0, 1, 1, 1, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0,
       0, 0, 0, 0, 1, 1, 0, 1, 1, 0, 1, 1, 1, 0, 0, 0, 0, 0, 0, 1, 1, 0,
       0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 1, 1, 0, 0,
       1, 0, 1, 1, 0, 0, 1, 1, 0, 0, 0, 1, 0, 0, 1, 1, 0, 1, 0, 1, 0, 1,
       0, 1, 0, 1, 0, 0, 0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0,
       1, 0, 1, 1, 1, 1, 0, 0, 0, 0, 1, 1, 0, 1, 0, 0, 1, 1, 0, 0, 0, 0,
       0, 0, 0, 1, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0, 1,
       1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1,
       1, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 1, 1, 1, 1,
       0, 1, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 1, 1, 0, 1, 1, 0, 0, 0,
       0, 1, 1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 1, 0, 1, 0,
       1, 0, 1, 1, 1, 1, 0, 1, 1, 0, 0, 1, 1, 0, 0, 1, 1, 0, 0, 0, 1, 1,
       0, 0, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 1,
       0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0,

In [68]:
print('Acurácia: %.2f%%' % (accuracy_score(y_treino, previsores_treino_logistica) * 100.0))

Acurácia: 76.38%


In [147]:
confusion_matrix(y_treino, previsores_treino_logistica)

array([[344,  49],
       [ 96, 125]])

### Validação Cruzada

In [70]:
# Separando os folds

kfold = KFold(n_splits = 30, shuffle=True, random_state = 5)

In [339]:
# Criando o modelo

modelo = LogisticRegression(random_state=1, max_iter=200, penalty='l2', tol=1, C=1, solver='lbfgs')
resultado = cross_val_score(modelo, previsores, alvo, cv = kfold)

/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_logistic.py:818: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG,


In [340]:
# Usando a média e o desvio padrão

print('Acurácia Média: %.2f%%' % (resultado.mean() * 100.0))

Acurácia Média: 76.79%


Naive Bayes = 78.57% (treino e teste) - 121 acertos e 74.71% (validação cruzada) - previsores

Máquinas de Vetores de Suporte (SVM) = 80.52% (Treino e Teste) - 124 acertos e 75.25% (validação cruzada) - previsores.

Regressão Logistica = 80.52% (treino e teste) - 124 acertos e 76.79% (validação cruzada) - previsores.

# APRENDIZADO BASEADA EM INSTÂNCIA (KNN)

### Treinamento Algoritmo

In [73]:
from sklearn.neighbors import KNeighborsClassifier

In [74]:
knn = KNeighborsClassifier(n_neighbors=15, metric='minkowski', p=1)
knn.fit(x_treino, y_treino)

KNeighborsClassifier(n_neighbors=15, p=1)

In [75]:
previsoes_knn = knn.predict(x_teste)
previsoes_knn

array([1, 0, 0, 1, 0, 0, 1, 1, 0, 0, 1, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 1, 1, 0, 0, 0, 0, 0, 0, 1,
       1, 0, 0, 0, 0, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 1, 1,
       1, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0,
       1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1,
       0, 0, 1, 0, 1, 1, 0, 1, 0, 1, 1, 0, 0, 0, 0, 0, 0, 1, 0, 1, 1, 0,
       0, 1, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0])

In [76]:
y_teste

array([1, 0, 0, 1, 0, 0, 1, 1, 0, 0, 1, 1, 0, 0, 0, 0, 1, 0, 0, 0, 1, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 1,
       1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 0, 0, 1, 1, 1, 1, 0, 0, 1, 1, 1,
       1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0,
       1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1,
       0, 1, 1, 1, 1, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0,
       0, 1, 0, 1, 0, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 1, 0, 0])

In [77]:
print('Acurácia: %.2f%%' % (accuracy_score(y_teste, previsoes_knn) * 100))

Acurácia: 81.17%


In [78]:
print(classification_report(y_teste, previsoes_knn))

              precision    recall  f1-score   support

           0       0.86      0.87      0.87       107
           1       0.70      0.68      0.69        47

    accuracy                           0.81       154
   macro avg       0.78      0.78      0.78       154
weighted avg       0.81      0.81      0.81       154



In [79]:
confusion_matrix(y_teste, previsoes_knn)

array([[93, 14],
       [15, 32]])

### Análise de dados de treino

In [80]:
previsoes_treino_knn = knn.predict(x_treino)
previsoes_treino_knn

array([1, 0, 1, 0, 1, 1, 1, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0,
       0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1,
       0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0,
       0, 0, 0, 0, 1, 1, 1, 0, 0, 0, 1, 1, 0, 0, 0, 1, 0, 1, 1, 1, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0,
       1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0,
       0, 0, 0, 1, 1, 0, 0, 1, 0, 1, 1, 0, 0, 0, 1, 1, 0, 0, 0, 1, 0, 0,
       0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 1, 1, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0,
       1, 1, 1, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 1, 0, 0, 1, 0, 1, 1, 1, 1,
       0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0,
       0, 1, 0, 0, 0, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 0, 0, 0, 1, 0, 1, 0,
       1, 0, 1, 1, 1, 1, 0, 1, 1, 0, 0, 0, 1, 0, 1, 1, 1, 1, 0, 0, 0, 1,
       1, 0, 0, 0, 1, 0, 0, 1, 0, 1, 0, 0, 0, 1, 0, 1, 1, 0, 1, 1, 0, 0,
       0, 0, 1, 0, 0, 1, 0, 1, 0, 1, 0, 0, 1, 1, 0,

In [81]:
print('Acurácia: %.2f%%' % (accuracy_score(y_treino, previsoes_treino_knn) * 100))

Acurácia: 77.85%


In [148]:
confusion_matrix(y_treino, previsoes_treino_knn)

array([[342,  51],
       [ 85, 136]])

### Validação Cruzada

In [82]:
# Separando os dados em folds

kfold = KFold(n_splits = 30, shuffle=True, random_state=5)

In [341]:
# Criando o modelo

modelo = KNeighborsClassifier(n_neighbors=15, metric='minkowski', p=1)
resultado = cross_val_score(modelo, previsores, alvo, cv = kfold)

In [342]:
# Usando a média e o desvio padrão

print('Acurácia Média: %.2f%%' % (resultado.mean() * 100.0))

Acurácia Média: 74.74%


Naive Bayes = 78.57% (treino e teste) - 121 acertos e 74.71% (validação cruzada) - previsores

Máquinas de Vetores de Suporte (SVM) = 80.52% (Treino e Teste) - 124 acertos e 75.25% (validação cruzada) - previsores.

Regressão Logistica = 80.52% (treino e teste) - 124 acertos e 76.79% (validação cruzada) - previsores.

Aprendiza Baseada em Instância (KNN) = 81.17% - 125 acertos e 74.74% (validação cruzada) - previsores.

# ÁRVORE DE DECISÃO

### Treinamento Algoritmo

In [85]:
from sklearn.tree import DecisionTreeClassifier

In [136]:
arvore = DecisionTreeClassifier(criterion='gini', random_state = 1, max_depth=4)
arvore.fit(x_treino, y_treino)

DecisionTreeClassifier(max_depth=4, random_state=1)

In [137]:
previsores_arvore = arvore.predict(x_teste)
previsores_arvore

array([1, 0, 0, 1, 0, 0, 1, 1, 0, 0, 1, 1, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0,
       0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 1, 1, 0, 0, 0, 1,
       1, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 0,
       1, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])

In [138]:
print('Acurácia: %.2f%%' % (accuracy_score(y_teste, previsores_arvore) * 100.0))

Acurácia: 76.62%


In [152]:
print(classification_report(y_teste, previsores_arvore))

              precision    recall  f1-score   support

           0       0.79      0.90      0.84       107
           1       0.67      0.47      0.55        47

    accuracy                           0.77       154
   macro avg       0.73      0.68      0.70       154
weighted avg       0.75      0.77      0.75       154



In [153]:
confusion_matrix(y_teste, previsores_arvore)

array([[96, 11],
       [25, 22]])

### Analise de Treino. 

In [141]:
previsoes_treino_arvore = knn.predict(x_treino)
previsoes_treino_arvore

array([1, 0, 1, 0, 1, 1, 1, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0,
       0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1,
       0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0,
       0, 0, 0, 0, 1, 1, 1, 0, 0, 0, 1, 1, 0, 0, 0, 1, 0, 1, 1, 1, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0,
       1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0,
       0, 0, 0, 1, 1, 0, 0, 1, 0, 1, 1, 0, 0, 0, 1, 1, 0, 0, 0, 1, 0, 0,
       0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 1, 1, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0,
       1, 1, 1, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 1, 0, 0, 1, 0, 1, 1, 1, 1,
       0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0,
       0, 1, 0, 0, 0, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 0, 0, 0, 1, 0, 1, 0,
       1, 0, 1, 1, 1, 1, 0, 1, 1, 0, 0, 0, 1, 0, 1, 1, 1, 1, 0, 0, 0, 1,
       1, 0, 0, 0, 1, 0, 0, 1, 0, 1, 0, 0, 0, 1, 0, 1, 1, 0, 1, 1, 0, 0,
       0, 0, 1, 0, 0, 1, 0, 1, 0, 1, 0, 0, 1, 1, 0,

In [144]:
print('Acurácia: %.2f%%' % (accuracy_score(y_treino, previsoes_treino_arvore) * 100))

Acurácia: 77.85%


In [146]:
confusion_matrix(y_treino, previsoes_treino_arvore)

array([[342,  51],
       [ 85, 136]])

### Validação Cruzada

In [149]:
# Separando os dados em folds

kfold = KFold(n_splits = 30, shuffle=True, random_state=5)

In [343]:
# Criando o modelo

modelo = DecisionTreeClassifier(criterion='gini', random_state = 1, max_depth=4)
resultado = cross_val_score(modelo, previsores, alvo, cv = kfold)

In [344]:
# Usando a média e o desvio padrão

print('Acurácia Média: %.2f%%' % (resultado.mean() * 100.0))

Acurácia Média: 71.71%


Naive Bayes = 78.57% (treino e teste) - 121 acertos e 74.71% (validação cruzada) - previsores

Máquinas de Vetores de Suporte (SVM) = 80.52% (Treino e Teste) - 124 acertos e 75.25% (validação cruzada) - previsores.

Regressão Logistica = 80.52% (treino e teste) - 124 acertos e 76.79% (validação cruzada) - previsores.

Aprendiza Baseada em Instância (KNN) = 81.17% - 125 acertos e 76.79% (validação cruzada) - previsores.

Árvore de Decisão = 76.62% - 118 acertos e 71.71% (validação cruzada) - previsores.

# RANDOM FOREST

### Treinamento Algoritmo

In [327]:
from sklearn.ensemble import RandomForestClassifier

In [328]:
random = RandomForestClassifier(n_estimators=184, criterion='entropy', random_state=6, max_depth=8)
random.fit(x_treino, y_treino)

RandomForestClassifier(criterion='entropy', max_depth=8, n_estimators=184,
                       random_state=6)

In [329]:
previsores_random = random.predict(x_teste)
previsores_random

array([1, 0, 0, 1, 0, 0, 1, 1, 0, 0, 1, 1, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0,
       0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 0, 0, 0, 1,
       1, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 1, 1, 0, 1, 1, 1,
       1, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 1,
       0, 1, 1, 1, 1, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 1, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0])

In [330]:
print('Acurácia: %.2f%%' %(accuracy_score(y_teste, previsores_random) * 100.0))

Acurácia: 84.42%


In [331]:
print(classification_report(y_teste, previsores_random))

              precision    recall  f1-score   support

           0       0.87      0.91      0.89       107
           1       0.77      0.70      0.73        47

    accuracy                           0.84       154
   macro avg       0.82      0.80      0.81       154
weighted avg       0.84      0.84      0.84       154



In [332]:
confusion_matrix(y_teste, previsores_random)

array([[97, 10],
       [14, 33]])

### Analise de Treino.

In [333]:
previsores_treino_random = random.predict(x_treino)
previsores_treino_random

array([1, 0, 0, 0, 1, 1, 1, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0,
       0, 0, 0, 0, 0, 1, 0, 1, 1, 0, 1, 1, 1, 0, 0, 0, 0, 0, 0, 1, 1, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0,
       0, 0, 1, 1, 0, 1, 1, 1, 0, 0, 0, 1, 0, 0, 1, 1, 0, 1, 0, 1, 0, 0,
       0, 1, 0, 1, 0, 0, 0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0,
       1, 0, 1, 1, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0,
       0, 0, 0, 1, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0, 1,
       1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 1, 0, 0, 1, 0, 0, 0, 0, 0, 1,
       1, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 1, 1, 1, 1,
       0, 1, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 1, 1, 0, 1, 1, 0, 0, 0,
       0, 1, 1, 0, 0, 1, 1, 0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 0, 1, 0, 1, 0,
       1, 0, 1, 1, 1, 1, 0, 1, 1, 0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 0, 0, 1,
       1, 0, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 1,
       0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0,

In [334]:
print('Acurácia: %.2f%%' %(accuracy_score(y_treino, previsores_treino_random) * 100.0))

Acurácia: 93.16%


In [210]:
confusion_matrix(y_treino, previsores_treino_random)

array([[381,  12],
       [ 23, 198]])

### Validação Cruzada

In [335]:
# Separando os dados em folds

kfold = KFold(n_splits = 30, shuffle=True, random_state=5)

In [345]:
# Criando o modelo

modelo = RandomForestClassifier(n_estimators=184, criterion='entropy', random_state=6, max_depth=8)
resultado = cross_val_score(modelo, previsores, alvo, cv = kfold)

In [346]:
# Usando a média e o desvio padrão

print('Acurácia Média: %.2f%%' % (resultado.mean() * 100.0))

Acurácia Média: 75.36%


Naive Bayes = 78.57% (treino e teste) - 121 acertos e 74.71% (validação cruzada) - previsores

Máquinas de Vetores de Suporte (SVM) = 80.52% (Treino e Teste) - 124 acertos e 75.25% (validação cruzada) - previsores.

Regressão Logistica = 80.52% (treino e teste) - 124 acertos e 76.79% (validação cruzada) - previsores.

Aprendiza Baseada em Instância (KNN) = 81.17% - 125 acertos e 76.79% (validação cruzada) - previsores.

Árvore de Decisão = 76.62% - 118 acertos e 71.71% (validação cruzada) - previsores.

Random Forast = 84.42% - 130 acertos e  75.36% (validação cruzada) - previsores.

# SALVANDO DADOS PARA DEPLOY

In [356]:
previsores

array([[  6.   , 148.   ,  72.   , ...,  33.6  ,   0.627,  50.   ],
       [  1.   ,  85.   ,  66.   , ...,  26.6  ,   0.351,  31.   ],
       [  8.   , 183.   ,  64.   , ...,  23.3  ,   0.672,  32.   ],
       ...,
       [  5.   , 121.   ,  72.   , ...,  26.2  ,   0.245,  30.   ],
       [  1.   , 126.   ,  60.   , ...,  30.1  ,   0.349,  47.   ],
       [  1.   ,  93.   ,  70.   , ...,  30.4  ,   0.315,  23.   ]])

In [357]:
alvo

array([1, 0, 1, 0, 1, 0, 1, 0, 1, 1, 0, 1, 0, 1, 1, 1, 1, 1, 0, 1, 0, 0,
       1, 1, 1, 1, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0, 0, 1,
       0, 1, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 1, 0,
       1, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0,
       1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1,
       1, 1, 0, 0, 1, 1, 1, 0, 0, 0, 1, 0, 0, 0, 1, 1, 0, 0, 1, 1, 1, 1,
       1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0,
       1, 1, 0, 0, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 1,
       0, 1, 0, 1, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 0, 0, 1, 1, 0, 1, 0, 1,
       1, 1, 0, 0, 0, 0, 0, 0, 1, 1, 0, 1, 0, 0, 0, 1, 1, 1, 1, 0, 1, 1,
       1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 0, 1, 1, 1, 1, 0, 0, 0,
       1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 1, 0, 1, 0, 0,
       1, 0, 1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 1, 1, 0,
       0, 1, 0, 0, 0, 1, 1, 1, 0, 0, 1, 0, 1, 0, 1,

In [359]:
np.savetxt('previsores.csv', previsores, delimiter=',')

np.savetxt('alvo.csv', alvo, delimiter=',')